slither/detectors/operations/cache_array_length.py

```diff
@@ -0,0 +1,215 @@
+        for (uint i = 0; i < array.length; i++)
+        {
+            // code that does not modify length of `array`
+        }
+    }
+}
+```
+Since the `for` loop in `f` doesn't modify `array.length`, it is more gas efficient to cache it in some local variable and use that variable instead, like in the following example:
+
+```solidity
+contract C
+{
+    uint[] array;
+    
+    function f() public 
+    {
+        uint array_length = array.length;
+        for (uint i = 0; i < array_length; i++)
+        {
+            // code that does not modify length of `array`
+        }
+    }
+}
+```
+    """
+    WIKI_RECOMMENDATION = (
+        "Cache the lengths of storage arrays if they are used and not modified in `for` loops."
+    )
+
+    @staticmethod
+    def _is_identifier_member_access_comparison(exp: BinaryOperation) -> bool:
+        """
+        Checks whether a BinaryOperation `exp` is an operation on Identifier and MemberAccess.
+        """
+        return (
+            isinstance(exp.expression_left, Identifier)
+            and isinstance(exp.expression_right, MemberAccess)
+        ) or (
+            isinstance(exp.expression_left, MemberAccess)
+            and isinstance(exp.expression_right, Identifier)
+        )
+
+    @staticmethod
+    def _extract_array_from_length_member_access(exp: MemberAccess) -> StateVariable:
+        """
+        Given a member access `exp`, it returns state array which `length` member is accessed through `exp`.
+        If array is not a state array or its `length` member is not referenced, it returns `None`.
+        """
+        if exp.member_name != "length":
+            return None
+        if not isinstance(exp.expression, Identifier):
+            return None
+        if not isinstance(exp.expression.value, StateVariable):
+            return None
+        if not isinstance(exp.expression.value.type, ArrayType):
+            return None
+        return exp.expression.value
+
+    @staticmethod
+    def _is_loop_referencing_array_length(
+        node: Node, visited: Set[Node], array: StateVariable, depth: int
+    ) -> True:
+        """
+        For a given loop, checks if it references `array.length` at some point.
+        Will also return True if `array.length` is referenced but not changed.
+        This may potentially generate false negatives in the detector, but it was done this way because:
+        - situations when array `length` is referenced but not modified in loop are rare
+        - checking if `array.length` is indeed modified would require much more work
+        """
+        visited.add(node)
+        if node.type == NodeType.STARTLOOP:
+            depth += 1
+        if node.type == NodeType.ENDLOOP:
+            depth -= 1
+        if depth == 0:
+            return False
+
+        # Array length may change in the following situations:
+        # - when `push` is called
+        # - when `pop` is called
+        # - when `delete` is called on the entire array
```

Comment:
I may handle that. Should we also consider internal calls in such a scenario?

Reply:
Good point. If you use `node.all_slithir_operations` instead of `node.irs`, it will include the operations of internal calls (https://crytic.github.io/slither/slither/core/cfg/node.html#Node.all_slithir_operations)